lib/std/once.zig

```diff
@@ -7,65 +7,120 @@ const std = @import("std.zig");
 const builtin = std.builtin;
 const testing = std.testing;
 
-pub fn once(comptime f: fn () void) Once(f) {
+pub fn once(comptime f: anytype) Once(f) {
     return Once(f){};
 }
 
 /// An object that executes the function `f` just once.
-pub fn Once(comptime f: fn () void) type {
+pub fn Once(comptime f: anytype) type {
+    const T = @TypeOf(f);
+    const ty_info = @typeInfo(T);
+    if (ty_info != .Fn)
+        @compileError("expected function type, found " ++ @typeName(T));
+
+    const fn_info = ty_info.Fn;
+    if (fn_info.is_generic)
+        @compileError("cannot instantiate Once with a generic function");
+    if (fn_info.is_var_args)
+        @compileError("cannot instantiate Once with a variadic function");
+
+    // The return type must be void or !void.
+    const Return = fn_info.return_type.?;
+    if (!switch (@typeInfo(Return)) {
+        .Void => true,
+        .ErrorUnion => |info| info.payload == void,
+        else => false,
+    }) {
+        @compileError("expected function returning void or !void, found " ++ @typeName(T));
+    }
+
+    const Args = std.meta.ArgsTuple(T);
+
+    // The field order minimizes the amount of space wasted to padding.
     return struct {
-        done: bool = false,
         mutex: std.Thread.Mutex = std.Thread.Mutex{},
+        done: std.atomic.Atomic(bool) = std.atomic.Atomic(bool).init(false),
+        err_val: Return = undefined,
 
-        /// Call the function `f`.
+        /// Call the function `f` with the supplied arguments, use .{} if the
+        /// function takes no arguments.
         /// If `call` is invoked multiple times `f` will be executed only the
-        /// first time.
+        /// first time. The arguments are unused for every call but the first
+        /// one.
+        /// If `f` returns an error the initialization is marked as done and,
+        /// for every following call, the same error value is returned.
         /// The invocations are thread-safe.
-        pub fn call(self: *@This()) void {
-            if (@atomicLoad(bool, &self.done, .Acquire))
-                return;
+        pub fn call(self: *@This(), args: Args) Return {
```

Comment:
Is there a way we could have this function return whether or not it performed the initialization?

Reply:
What's the use case for that?

Reply:
It really feels like there should be one but can't think of any ATM.

Reply:
In the example of WSAStartup, if this is used as a "fallback initialization" routine, then if the call returns that it has already been initialized, then the function can fail right away rather than try to call the function that failed again, i.e.


```zig
fn someWsaFunc() {
    callTheFunc() catch |e| switch (e) {
        error.WsaNotInitialized => {
            if (!wsaStartupOnce.call(2, 2)) {
                unreachable,
            }
            return someWsaFunc();
        }
    };
}

```

Reply:
> then the function can fail right away rather than try to call the function that failed again

The function is called once, no matter if an error is returned or not. There's no way to disarm the `Once` object (it wouldn't be called `once` then) and the error condition is propagated to every caller, be it from another thread or at another time.

Providing a way to check if the function was called is possible but the idea here is to call `call` and call it a day (heh), the check is lightweight and avoids any possible data race.

Reply:
@LemonBoy, sorry my message was not clear.  The caller can try to call the function that failed again, not the "once function" again.  I think the code example I shared explains the use case.

Reply:
I still don't understand what your point is. What is the function that fails?

Reply:
`callTheFunc` is a WSA function that fails.  We've decided that is a WSA function fails, we will call WSAStartup and then try calling it again. In this case, we use `std.once` to call wsastartup, so if we detect that it's already been called, then we forego calling the function again.

Reply:
That doesn't make sense, you should call `call()` anyway and forget about it. That's the whole point of `once`.

Reply:
We are calling it anyway.  Again, I'm not saying we're not calling `call`, I'm saying after we call `call`, if we know whether or not we initialized, then we know whether or not to call the function that failed again or just to fail immediately. Look at the code example above.

Reply:
I see. My offer here is to add an `wasCalled` predicate (or with a better name).

Reply:
Feels like a race condition.

Reply:
But now that I think about it, the whole use case I've presented here is also a race condition :)

Reply:
I'm fine committing as is and waiting for a real use case before adding this.

Reply:
> Feels like a race condition.

Nope. The predicate returns `true` if the function was called and that's it, whenever you see it returning `false` you can't make any assumption about the state of the `Once` instance, that's what makes this kind of predicate tricky to use.

> I'm fine committing as is and waiting for a real use case before adding this.

Thank you, kind sir.

Reply:
> Nope. The predicate returns true if the function was called and that's it, whenever you see it returning false you can't make any assumption about the state of the Once instance, that's what makes this kind of predicate tricky to use.

The race condition would be.

1. call `wasCalled` and it returns false
2. another thread calls `call`
3. you call `call` and assume incorrectly that you are the one that got to call it

So if there's some sort of behavior that needs to know whether or not their `call` is the one that "gets through", then there's no way to know for sure.  Any thread needing this behavior would be introducing a "race condition".

Reply:
I'll quote my previous message again

> The predicate returns true if the function was called and that's it, whenever you see it returning false you can't make any assumption about the state of the Once instance, that's what makes this kind of predicate tricky to use.

`wasCalled` is only useful for determining if the function was run before the call returned and as such cannot be used to implement the `if (!wasCalled) call()` pattern (that'd defeat the whole point of using `Once`).

Please come back armed with willingness to read people's answers and a valid use case, if you find one.